Make correction factors text easier to understand (#1177)
riscv/riscv-cheri: src/cheri/rvy64-encoding.adoc

* `B[MW - 1:EW / 2]` is taken directly from the capability metadata.
* `B[EW / 2 - 1:0]`  is as specified in the pseudo-code above.

The design rationale for the two `EF` cases is as follows:

1. `EF = 1`: The exponent is 0. When `enableL8=1`, `L~8~` encodes the MSB of the length,

> which can be used to derive `T[MW-1:MW-2]`, forming a full MW-wide T field.

Reviewer: ```suggestion
which can be used to derive `T[MW-1:MW-2]`, forming a full MW-wide `T` field.
```

2. `EF = 0`: The exponent is _internal_ with E stored in the lower bits of `T` and
`B`, with `L~8~` used for the MSB of `E` when `enableL8=1`. `E` is chosen so that the most
significant non-zero bit of the length of the region aligns with `T[MW - 2]` such
that this bit is implied by `E`.

> To save space in the encoding metadata, the top two bits of T are not stored directly but reconstituted as follows:

Reviewer: ```suggestion
To save space in the encoding metadata, the top two bits of `T` are not stored directly but reconstituted as follows:
```

In the example shown in <<capformat_correction_illustration>>, the _top_ bound and the capability address lie in the same _s_-aligned block, but the _base_ bound doesn't, it is in the block below.
Therefore _c~t~_ is zero, but _c~b~_ is set to -1 to move the _base_ bound down into the _s_-aligned block below (i.e. from space~U~ to space~L~).

Therefore adding the _correction factor_ moves the computed bound into the correct _s_-aligned block.
Each _correction factor_ can have values of -1, 0 or +1, so that it is possible to move computed bound up or down by one _s_-aligned block.

[#capformat_correction_illustration]
.Example capability bounds that use the correction. The _base_ is an s-aligned block below the address, meaning the upper bits (above the mantissa) of the _base_ are smaller than the upper bits of the address. Therefore _c~b~_=-1. The address and _top_ are in the same region, so _c~t~_=0.
image::../cheri/img/capformat-correction-illustration.png[width=80%,align=center]

====== Correction Factor Calculations

`R` marks the lower boundary of the <<section_cap_representable_check,representable range>> (shown as r~base~ in <<capformat_correction_illustration>>).

> Comparing `A` (the bit range of the capability address which gives the offset within the 2^E+MW^-aligned block), `B` and `T` against `R` tells us whether each lies in the 2^E+MW^-aligned block containing `R` (`space~L~` in <<cap_bounds_map>>), the block above it (`space~U~`) or the block below (not shown in <<capformat_correction_illustration>>, all possible combinations are shown in <<capformat_correction_cases>>).

Reviewer 2: I don't think this makes sense as is. There is no address block that "contains R": it's only an MW-bit thing and there are shadows of it in all the regions. The point is you have MW-width B and T in [0, 2^MW), and you want to map them to [R, 2^{MW-1}) (the upper portion) in a low region, and [0, R) (the lower portion) in a high region. Pondering suggested replacement text now.

> Comparing `A` (the bit range of the capability address which gives the offset within the 2^E+MW^-aligned block), `B` and `T` against `R` tells us whether each lies in the 2^E+MW^-aligned block containing `R` (`space~L~` in <<cap_bounds_map>>), the block above it (`space~U~`) or the block below (not shown in <<capformat_correction_illustration>>, all possible combinations are shown in <<capformat_correction_cases>>).

Reviewer 2: ```suggestion
If `B`, `T` or `A` is less than `R`, it is inferred that they lie in the space above the s-alignment boundary straddled by the representable region, labeled space~U~ in xref:capformat_correction_illustration[xrefstyle=short].
```

----
A[MW-1:0] = a[E + MW - 1:E]; // offset into s-aligned region from capability address field
R[MW-1:0] = B - 2^MW-2^;        // r-base
----

NOTE: The comparisons in xref:cap_encoding_corrections[xrefstyle=short] are _unsigned_.

The correction factors _c~b~_ and _c~t~_ are calculated as follows.

NOTE: Mathematically, `B < R` requires `B[MW-1:MW-2] = 0`.

.Calculation of the _top_ and _base_ bound corrections
[#cap_encoding_corrections,options=header,cols="^1,^1,^1,^2",width="60%",align="center"]
|==============================================================================
| A < R      | T < R | B < R | Correction factors
.4+.^| false | false | false | _c~t~_=0, _c~b~_=0
             | false | true  | Impossible^1^
             | true  | false | _c~t~_=+1, _c~b~_=0
             | true  | true  | _c~t~_=+1, _c~b~_=+1
.4+.^| true  | false | false | _c~t~_=-1, _c~b~_=-1
             | false | true  | Impossible^1^
             | true  | false | _c~t~_=0, _c~b~_=-1
             | true  | true  | _c~t~_=0, _c~b~_=0
|==============================================================================

^1^ Since `B[MW-1:MW-2] = 0` in these cases, `R[MW-1:MW-2] = 3`. +

> As a result `T[MW-1:MW-2]` is at most 2 (according to this <<formula_top_bits_of_t,formula>>). +

Reviewer: ```suggestion
^1^ Since `B[MW-1:MW-2] = 0` in these cases, `R[MW-1:MW-2] = 3` as the value has wrapped. +
As a result `T[MW-1:MW-2]` is at most 2 if LMSB and LCout are both 1 (according to this <<formula_top_bits_of_t,formula>>). +
```

include ranges which may not be present when the bounds are decoded:

* If `E = 0` the lower section does not exist.
* If `E+MW=XLEN` then the top section is only the least significant bit of
 _c~t~_ for the _top_ bound, and top section does not exist for the _base_ bound.
* If `E+MW>XLEN` then neither top section exists, and so the bounds are calculated
 with no dependency on the address field _a_.

The compressed bounds encoding allows the address to roam over the _representable range_, (`space~R~` in <<cap_bounds_map>> and <<capformat_correction_illustration>>), while maintaining the original bounds.
The overall effect is that the address can roam at least `2^E+MW^/4` bytes below the _base_ bound and at least `2^E+MW^/4` bytes above the _top_ bound while still allowing the bounds to be correctly decoded.

[#section_cap_bounds_correction_factors, reftext="Correction Factors"]
===== Correction Factors

Substituting `B`/`T` into the capability address, as described in xref:section_cap_bounds_format[xrefstyle=short] above, only decodes to the correct _base_ or _top_ bound if `B`/`T` and the capability address both lie in the _same_ _s_=2^E+MW^-aligned block.

> In this case the relevant _correction factor_ is zero (_c~b~_ for the _base_ and _c~t~_ for the _top_).

Author: We haven't explained what a correction factor is at this point, I'd just drop this sentence.

Reviewer: reworded - will commit shortly.

[#section_cap_bounds_correction_factors]
===== Correction Factors

As described in xref:section_cap_bounds_format[xrefstyle=short] below, the base, _b_, and top, _t_, addresses are computed by substituting `B` and `T` for the upper `MW` bits of the capability address, leaving the address's own 2^E+MW^-aligned block unchanged.
This substitution alone is only correct if `B`/`T` and the capability address lie in the same 2^E+MW^-aligned block.

> When they do not (address 0x20000 in <<capformat_correction_illustration>>), the correction factors c~t~ and c~b~ shift the substituted block up or down by one so that the computed bound falls in the block that actually contains `b`/`t`.

Reviewer: ```suggestion
When they do not (address 0x20000 in <<capformat_correction_illustration>>), the correction factors c~t~ and c~b~ shift the substituted block up or down by one so that the computed bound falls in the block that actually contains _b_ or _t_.
```

`R` marks the lower boundary of the <<section_cap_representable_check,representable range>> (see <<cap_bounds_map>>).

> Comparing `A`, `T` and `B` against `R` tells us whether each lies in the 2^E+MW^-aligned block containing `R` (`space~L~` in <<cap_bounds_map>>) or the block above it (`space~U~`).

Reviewer: ```suggestion
Comparing `A` (the bit range of the capability address which gives the offset within the 2^E+MW^-aligned block), `B` and `T` against `R` tells us whether each lies in the 2^E+MW^-aligned block containing `R` (`space~L~` in <<cap_bounds_map>>) or the block above it (`space~U~`).
```

> As shown in xref:section_cap_bounds_format[xrefstyle=short] below, the upper bits of the bound are computed by substituting B or T into the upper bits of the capability address.

Reviewer: ```suggestion
As shown in xref:section_cap_bounds_format[xrefstyle=short] below, the upper bits of the bound are computed by substituting `B` or `T` into the upper bits of the capability address.
```

[#section_cap_bounds_correction_factors]
===== Correction Factors

As described in xref:section_cap_bounds_format[xrefstyle=short] below, the `_base_` and `_top_` addresses are computed by substituting `B` and `T` for the upper `MW` bits of the capability address, leaving the address's own 2^E+MW^-aligned block unchanged.
This substitution is only correct if `B`/`T` and the capability address lie in the same 2^E+MW^-aligned block.
When they do not (as with the base in <<capformat_correction_illustration>>), the correction factors c~t~ and c~b~ shift the substituted block up or down by one so that the computed bound falls in the block that actually contains `_base_` or `_top_`.

> `R` marks the lower boundary of the <<section_cap_representable_check,representable range>> (see <<cap_bounds_map>>).

Reviewer: isn't this r_base in the diagram? Also no reason to refer back to figure 2 - it's also on the local figure.

Author: Oh yes that text also needs updating

Reviewer: I've fixed this.

[#section_cap_bounds_correction_factors]
===== Correction Factors

> As described in xref:section_cap_bounds_format[xrefstyle=short] below, the base, _b_, and top, _t_, addresses are computed by substituting `B` and `T` for the upper `MW` bits of the capability address, leaving the address's own 2^E+MW^-aligned block unchanged.

Reviewer: >for the upper `MW` bits of the capability address

this isn't true - they go into bit positions [E+MW-1 : E]

Reviewer: shouldn't this be something like:

The upper bits of the capablity address determine the  2^E+MW^-aligned block, and then the values of `B` and `T` are substituted into the address to form the bounds.

Reviewer: I've just realised that this incorrect text was on line 327 before the edit so it's not a new problem.

Reviewer: filed as #1185

This substitution alone is only correct if `B`/`T` and the capability address lie in the same 2^E+MW^-aligned block.

> When they do not (address 0x20000 in <<capformat_correction_illustration>>), the correction factors c~t~ and c~b~ shift the substituted block up or down by one so that the computed bound falls in the block that actually contains _b_ or _t_.

Reviewer: why address 0x20000? that's right on the boundary and isn't the value used for either b or t, which not pick an address which is actually used?

Reviewer: see #1184 the diagram is quite weird

Reviewer: this is an exsiting problem - not related to this PR

`R` marks the lower boundary of the <<section_cap_representable_check,representable range>> (see <<cap_bounds_map>>).

> Comparing `A` (the bit range of the capability address which gives the offset within the 2^E+MW^-aligned block), `B` and `T` against `R` tells us whether each lies in the 2^E+MW^-aligned block containing `R` (`space~L~` in <<cap_bounds_map>>) or the block above it (`space~U~`).

Reviewer: or also the block below it as correction factors can be +1/0/-1 moving the block up or down

> As shown in xref:section_cap_bounds_format[xrefstyle=short] below, the upper bits of the bound are computed by substituting `B` or `T` into the upper bits of the capability address.

Reviewer: this is almost a repeat of line 312?

> This substitution is only correct if the memory bound and the capability address lie within the same 2^E+MW^-aligned region.

Reviewer: this is also repeated

> When they do not (address 0x20000 in <<capformat_correction_illustration>>), the correction factors c~t~ and c~b~ adjust the computed upper bits to select the correct region.

Reviewer: repeated again

`A` and `R` are calculated as follows.

[source,subs="quotes"]
----
A[MW-1:0] = a[E + MW - 1:E]; // capability address field
R[MW-1:0] = B - 2^MW-2^
----

NOTE: The comparisons in xref:cap_encoding_corrections[xrefstyle=short] are _unsigned_.

The correction factors c~b~ and c~t~ are calculated as follows.

.Calculation of the top and base address corrections
[#cap_encoding_corrections,options=header,cols="^1,^1,^1,^2",width="60%",align="center"]
|==============================================================================
| A < R      | T < R | B < R | Correction factors
.4+.^| false | false | false | _c~t~_=0, _c~b~_=0
             | false | true  | Impossible^1^
             | true  | false | _c~t~_=+1, _c~b~_=0
             | true  | true  | _c~t~_=+1, _c~b~_=+1
.4+.^| true  | false | false | _c~t~_=-1, _c~b~_=-1
             | false | true  | Impossible^1^
             | true  | false | _c~t~_=0, _c~b~_=-1
             | true  | true  | _c~t~_=0, _c~b~_=0
|==============================================================================

^1^ `B < R` requires `B[MW-1:MW-2] = 0`, which makes `R[MW-1:MW-2] = 3`.

> Since `T[MW-1:MW-2] = B[MW-1:MW-2] + LCout + LMSB` (xref:section_cap_bounds_decoding[xrefstyle=short]) is at most 2 in that case, `T < R` always holds when `B < R`, ruling out c~t~=-1,c~b~=0 and c~t~=0,c~b~=+1.

Reviewer: I think that this needs some human-style description, I'm sure the text is correct but it's difficult to follow in this form.

Author: I can just simplify it to say `B<R` implies `T<R` without these details?

Reviewer: I've made it easier to read.

* If `E+MW>XLEN` then neither top section exists, and so the bounds are calculated
 with no dependency on the address field _a_.

The compressed bounds encoding allows the address to roam over the _representable_ range, `space~R~`, while maintaining the original bounds  (see xref:section_cap_bounds_correction_factors[xrefstyle=short] above for how
`R` and the corrections are calculated).

> The overall effect is that the  address can roam `2^E+MW^/4` bytes below the base address and at least  `2^E+MW^/4` bytes above the top address while still allowing the bounds to be

Reviewer: ```suggestion
The overall effect is that the  address can roam at least `2^E+MW^/4` bytes below the base address and at least  `2^E+MW^/4` bytes above the top address while still allowing the bounds to be
```
